### PR TITLE
New ingest module

### PR DESCRIPTION
**Pull request by mullr**

### CODE REVIEW

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 1]
```diff
@@ -0,0 +1,1065 @@
...
+//! use auxon_sdk::plugin_utils::ingest::{Config, init_tracing};
+//!
+//! #[derive(Serialize, Deserialize, Clone, Debug, Default)]
+//! pub struct MyConfig {
+//!     # This can be set with the MY_PLUGIN_SETTING environment variable
+//!     pub setting: Option<u32>,
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() -> Box<dyn std::error::Error> {
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:57]: Probably worth adding an example to the examples dir for this too.

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 2]
```diff
@@ -0,0 +1,1065 @@
...
+    /// structure represents the fully reconcicled configuration.
+    ///
+    /// * `env_prefix`: The prefix used for environment variable based
+    ///   settings for members of the configuration struct (type
+    ///   param `T`).
+    pub fn load(env_prefix: &str) -> Result<Config<T>, Box<dyn std::error::Error>> {
+        Self::load_custom(env_prefix, |_, _| None)
+    }
+
+    /// Load configuration, like `Config::load`, but allows passing a
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:143]: Should bracket `[]` any references to types/etc defined in this crate so rustdoc will produce links.

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 3]
```diff
@@ -0,0 +1,1065 @@
...
+    /// `map_env_val` hook.  This can be used to implement
+    /// non-standard environment deserialization, for value types
+    /// which aren't correctly handled by the `envy` crate.
+    ///
+    /// * `map_env_val`: A function which will be called for every
+    ///   environment variable. If it returns `Some((key, toml_value))`,
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:149]: It's not clear what the two str arguments to `map_env_val` are expected to be, would be helpful to have that called out in the docs.

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 4]
```diff
@@ -0,0 +1,1065 @@
...
+                                // plugins.ingest.collectors.<stem filename>
+                                .or_else(|| ingest.collectors.get(file_stem.alias()))
+                        } else if file_stem.looks_like_importer() {
+                            // plugins.ingest.collectors.<full bin filename>
+                            ingest
+                                .importers
+                                .get(file_stem.as_str())
+                                // plugins.ingest.collectors.<stem filename>
+                                .or_else(|| ingest.importers.get(file_stem.alias()))
+                        } else {
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:196]: Can we add mutator support too?

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 5]
```diff
@@ -0,0 +1,1065 @@
...
+        })
+    }
+
+    /// Connect to the configured Modality backend for ingest,
+    /// authenticate, and return a high-level ingest client.
+    pub async fn connect_and_authenticate(&self) -> Result<Client, Box<dyn std::error::Error>> {
+        let protocol_parent_url = if let Some(url) = &self.ingest.protocol_parent_url {
+            url.clone()
+        } else {
+            Url::parse("modality-ingest://127.0.0.1")?
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:259]: Shouldn't we be checking for `MODALITY_HOST` first?

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 6]
```diff
@@ -0,0 +1,1065 @@
...
+    ///   <div class="warning">Avoid sending duplicate `ordering` values for the same timeline.</div>
+    ///
+    /// * `attrs`: The attributes to attach to the event. While you can use this with anything
+    ///   that implements `IntoIterator`, it's idiomatic to use a literal slice, and to
+    ///   use `into()` to convert values to `AttrVal`:
+    ///   `client.send_event("ev", [("attr1", 42.into())]).await?;`
+    ///
+    ///   * These keys are automatically normalized, so you prepending "event." is optional.
+    ///
+    ///   * If "timestamp" or "event.timestamp" is not given here, the
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:550]: I think this is useful behavior to have, but worth noting that some plugins will intentionally not have timestamps and we may want to make an opt-in/out switch for this.

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 7]
```diff
@@ -0,0 +1,1065 @@
...
+        env::remove_var("MODALITY_REFLECTOR_CONFIG");
+        env::remove_var("MODALITY_REFLECTOR_ADDITIONAL_TIMELINE_ATTRIBUTES");
+        env::remove_var("MODALITY_REFLECTOR_OVERRIDE_TIMELINE_ATTRIBUTES");
+        env::remove_var("MODALITY_REFLECTOR_PROTOCOL_PARENT_URL");
+        env::remove_var("MODALITY_ALLOW_INSECURE_TLS");
+        env::remove_var("TEST_VAL");
+    }
+}
+
+// Reflector protocol parent url from the environment overrides config file
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/ingest.rs:1028]: Looks like this stuff can be removed?

**Review by jonlamb-gh — CHANGES_REQUESTED** [review]
A few minor things, otherwise LGTM.

File: client-libraries/rust/src/plugin_utils/ingest.rs [hunk 8]
```diff
@@ -0,0 +1,1065 @@
+//! A one-stop shop for most collector and import
```
[the remark this answers is not in the document]
Comment (mullr) [reply client-libraries/rust/src/plugin_utils/ingest.rs:1]: I was thinking we'd make a separate 'mutator' module for that purpose.

File: client-libraries/rust/src/plugin_utils/ingest.rs (hunk 4 above)
Comment (mullr) [reply client-libraries/rust/src/plugin_utils/ingest.rs:209]: Yes, but probably separately.

File: client-libraries/rust/src/plugin_utils/ingest.rs (hunk 5 above)
Comment (mullr) [reply client-libraries/rust/src/plugin_utils/ingest.rs:272]: That happens in `override_ingest_config_from_env`, which happens before this, so `self.ingest.protocol_parent_url` contains the final answer.

**Review by jonlamb-gh — APPROVED** [review]

**Pull request merged**